src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,438 @@
+#include "sol-mainloop-zephyr.h"
+#include "sol-socket-impl.h"
+#include "sol-vector.h"
+#include "sol-network-util.h"
+
+struct sol_socket_zephyr {
+    struct sol_socket base;
+
+    struct sol_list bound_list;
+
+    struct {
+        bool (*cb)(void *data, struct sol_socket *s);
+        const void *data;
+    } read, write;
+
+    struct sol_timeout *write_timeout;
+
+    struct net_context *unicast_context;
+    struct sol_ptr_vector mcast_contexts;
+    struct sol_ptr_vector bufs;
+
+    struct nano_sem lock;
+    bool read_available;
+    bool deleted;
+};
+
+#define RECV_STACKSIZE 1024
+static char __noinit __stack socket_recv_fiber_stack[RECV_STACKSIZE];
+static bool socket_recv_fiber_started;
+
+static struct sol_list bound_sockets = SOL_LIST_INIT(bound_sockets);
+static struct nano_sem bound_sockets_lock;
+
+static void
+socket_read_available(void *data)
+{
+    struct sol_socket_zephyr *s = data;
+    uint16_t count;
+
+    nano_task_sem_take(&s->lock, TICKS_UNLIMITED);
+    count = sol_ptr_vector_get_len(&s->bufs);
+    s->read_available = false;
+    nano_task_sem_give(&s->lock);
+
+    if (s->read.cb)
+        while (count--)
+            s->read.cb((void *)s->read.data, &s->base);
+}
+
+static void
+socket_signal_mainloop(struct sol_socket_zephyr *s, struct net_buf *buf)
+{
+    struct mainloop_event me = {
+        .cb = socket_read_available,
+        .data = s
+    };
+    int ret;
+
+    nano_fiber_sem_take(&s->lock, TICKS_UNLIMITED);
+
+    ret = sol_ptr_vector_append(&s->bufs, buf);
+    SOL_INT_CHECK_GOTO(ret, < 0, err);
+
+    if (!s->read_available) {
+        sol_mainloop_event_post(&me);
+        s->read_available = true;
+    }
+
+err:
+    nano_fiber_sem_give(&s->lock);
+}
+
+static void
+socket_del(struct sol_socket_zephyr *s)
+{
+    struct net_context *ctx;
+    struct net_buf *buf;
+    uint16_t idx;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&s->bufs, buf, idx)
+        ip_buf_unref(buf);
+    sol_ptr_vector_clear(&s->bufs);
+
+    net_context_put(s->unicast_context);
+    SOL_PTR_VECTOR_FOREACH_IDX (&s->mcast_contexts, ctx, idx)
+        net_context_put(ctx);
+    sol_ptr_vector_clear(&s->mcast_contexts);
+
+    free(s);
+}
+
+static void
+socket_recv_fiber(int arg1, int arg2)
+{
+    struct sol_list *itr = &bound_sockets;
+
+    while (1) {
+        struct sol_socket_zephyr *s;
+        struct net_context *ctx;
+        struct net_buf *buf;
+        uint16_t i;
+
+        nano_fiber_sem_take(&bound_sockets_lock, TICKS_UNLIMITED);
+        if (sol_list_is_empty(&bound_sockets))
+            break;
+        if (itr == &bound_sockets)
+            itr = itr->next;
+        s = SOL_LIST_GET_CONTAINER(itr, struct sol_socket_zephyr, bound_list);
+        nano_fiber_sem_give(&bound_sockets_lock);
+
+        buf = net_receive(s->unicast_context, 100);
+        if (buf)
+            socket_signal_mainloop(s, buf);
+        SOL_PTR_VECTOR_FOREACH_IDX (&s->mcast_contexts, ctx, i) {
+            buf = net_receive(ctx, 100);
```

Comment:
On Tue, 29 Mar 2016 14:26:01 -0700, Vinicius Gomes wrote:
> > +        struct net_buf *buf;
> > +        uint16_t i;
> > +
> > +        nano_fiber_sem_take(&bound_sockets_lock, TICKS_UNLIMITED);
> > +        if (sol_list_is_empty(&bound_sockets))
> > +            break;
> > +        if (itr == &bound_sockets)
> > +            itr = itr->next;
> > +        s = SOL_LIST_GET_CONTAINER(itr, struct sol_socket_zephyr, bound_list);
> > +        nano_fiber_sem_give(&bound_sockets_lock);
> > +
> > +        buf = net_receive(s->unicast_context, 100);
> > +        if (buf)
> > +            socket_signal_mainloop(s, buf);
> > +        SOL_PTR_VECTOR_FOREACH_IDX (&s->mcast_contexts, ctx, i) {
> > +            buf = net_receive(ctx, 100);
> 
> For now I am more worried about wakeups, did you think about having a fiber per context?

I can try that. I don't expect to have many contexts here anyway.